File: server/Simulation/Net/ShipConfigurationPacket.cs
using System;
using System.Collections.Generic;
using server.ECS;

namespace server.Simulation.Net;

public struct ShipPart
{
    public sbyte GridX; // Changed to signed byte for negative coordinates
    public sbyte GridY; // Changed to signed byte for negative coordinates
    public byte Type; // 0=hull, 1=shield, 2=engine
    public byte Shape; // 1=triangle, 2=square
    public byte Rotation; // 0=0°, 1=90°, 2=180°, 3=270°

    public ShipPart(sbyte gridX, sbyte gridY, byte type, byte shape, byte rotation)
    {
        GridX = gridX;
        GridY = gridY;
        Type = type;
        Shape = shape;
        Rotation = rotation;
    }
}

public class ShipConfigurationPacket
{
    public NTT NTT { get; set; }
    public List<ShipPart> Parts { get; set; }

    public ShipConfigurationPacket()
    {
        Parts = new List<ShipPart>();
    }

    public static ShipConfigurationPacket FromBuffer(ReadOnlyMemory<byte> buffer)
    {
        var reader = new PacketReader(buffer);
        var packet = new ShipConfigurationPacket();

        packet.NTT = reader.ReadNtt();
        var partCount = reader.ReadInt16();

        for (int i = 0; i < partCount; i++)
        {
            var part = new ShipPart(
                reader.ReadSByte(), // gridX (signed byte)
                reader.ReadSByte(), // gridY (signed byte)
                reader.ReadByte(), // type
                reader.ReadByte(), // shape
                reader.ReadByte()  // rotation
            );
            packet.Parts.Add(part);
        }

        return packet;
    }
}

File: pixiejsClient/src/app/network/packets/ShipConfigurationPacket.ts
import { EvPacketWriter } from "../EvPacketWriter";
import { EvPacketReader } from "../EvPacketReader";
import { PacketHeader } from "../PacketHeader";
import { PacketId } from "../PacketHandler";

export interface ShipPart {
  gridX: number;
  gridY: number;
  type: number; // 0=hull, 1=shield, 2=engine
  shape: number; // 1=triangle, 2=square
  rotation: number; // 0=0°, 1=90°, 2=180°, 3=270°
}

export class ShipConfigurationPacket {
  header: PacketHeader;
  playerId: string;
  parts: ShipPart[];

  constructor(header: PacketHeader, playerId: string, parts: ShipPart[]) {
    this.header = header;
    this.playerId = playerId;
    this.parts = parts;
  }

  static create(playerId: string, parts: ShipPart[]): ArrayBuffer {
    const writer = new EvPacketWriter(PacketId.ShipConfiguration);
    writer.Guid(playerId);
    writer.i16(parts.length);

    for (const part of parts) {
      writer.i8(part.gridX);
      writer.i8(part.gridY);
      writer.i8(part.type);
      writer.i8(part.shape);
      writer.i8(part.rotation);
    }

    writer.FinishPacket();
    return writer.ToArray();
  }

  static fromBuffer(buffer: ArrayBuffer): ShipConfigurationPacket {
    const reader = new EvPacketReader(buffer);
    const header = reader.Header();
    const playerId = reader.Guid();
    const partCount = reader.i16();

    const parts: ShipPart[] = [];
    for (let i = 0; i < partCount; i++) {
      parts.push({
        gridX: reader.i8(),
        gridY: reader.i8(),
        type: reader.i8(),
        shape: reader.i8(),
        rotation: reader.i8(),
      });
    }

    return new ShipConfigurationPacket(header, playerId, parts);
  }
}
